packages/@blockprotocol/core/src/core-handler.ts

```diff
@@ -0,0 +1,345 @@
+import { v4 as uuid } from "uuid";
+
+import { ServiceHandler } from "./service-handler";
+import {
+  GenericMessageCallback,
+  Message,
+  MessageCallbacksByService,
+  MessageData,
+  PromiseRejecter,
+  PromiseResolver,
+  ResponseSettlersByRequestIdMap,
+  SendMessageArgs,
+} from "./types";
+
+/**
+ * Implements the Block Protocol Core Specification, i.e.:
+ * - listening for and dispatching messages via events
+ * - registering callbacks from service handlers
+ * - processing incoming messages and calling callbacks as appropriate
+ *
+ * This base class handles those common functions. There are two implementers,
+ * Block and Embedder, to handle the divergent functionality (e.g. init messages).
+ */
+export abstract class CoreHandler {
+  /** a default callback for use where no callback is registered for a message */
+  private readonly defaultMessageCallback?: GenericMessageCallback;
+
+  /** an object containing registered callbacks for messages, by service and message name */
+  private readonly messageCallbacksByService: MessageCallbacksByService;
```

Comment:
It started out as an object of objects, but I then converted it to an object of maps to avoid a theoretical vulnerability if someone allows someone else to pass in malicious message names.

The root thing could be a map as well, I just didn't convert it.